hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/request/validation/ValidatorRegistry.java

```diff
@@ -0,0 +1,278 @@
+ * a service.
+ */
+public class ValidatorRegistry<RequestType extends Enum<RequestType>> {
+
+  private final Map<Class<? extends Version>, EnumMap<RequestType,
+      EnumMap<RequestProcessingPhase, IndexedItems<Method, Integer>>>> indexedValidatorMap;
+
+  /**
+   * Creates a {@link ValidatorRegistry} instance that discovers validation
+   * methods in the provided package and the packages in the same resource.
+   * A validation method is recognized by all the annotations classes which
+   * are annotated by {@link RegisterValidator} annotation that contains
+   * important information about how and when to use the validator.
+   * @param validatorPackage the main package inside which validatiors should
+   *                         be discovered.
+   */
+  public ValidatorRegistry(Class<RequestType> requestType,
+                    String validatorPackage,
+                    Set<Class<? extends Version>> allowedVersionTypes,
+                    Set<RequestProcessingPhase> allowedProcessingPhases) {
+    this(requestType, ClasspathHelper.forPackage(validatorPackage), allowedVersionTypes, allowedProcessingPhases);
+  }
+
+  private Class<?> getReturnTypeOfAnnotationMethod(Class<? extends Annotation> clzz, String methodName) {
+    try {
+      return clzz.getMethod(methodName).getReturnType();
+    } catch (NoSuchMethodException e) {
+      throw new IllegalArgumentException("Method " + methodName + " not found in class:" + clzz.getCanonicalName());
+    }
+  }
+
+  /**
+   * Creates a {@link ValidatorRegistry} instance that discovers validation
+   * methods under the provided URL.
+   * A validation method is recognized by all annotations having the annotated by {@link RegisterValidator}
+   * annotation that contains important information about how and when to use
+   * the validator.
+   * @param searchUrls the path in which the annotated methods are searched.
+   */
+  public ValidatorRegistry(Class<RequestType> requestType,
+                    Collection<URL> searchUrls,
+                    Set<Class<? extends Version>> allowedVersionTypes,
+                    Set<RequestProcessingPhase> allowedProcessingPhases) {
+    Set<Class<? extends Annotation>> validatorsToBeRegistered =
+        new Reflections(new ConfigurationBuilder().setUrls(ClasspathHelper.forPackage(""))
+            .setScanners(Scanners.TypesAnnotated)
+            .setParallel(true)).getTypesAnnotatedWith(RegisterValidator.class).stream()
+            .filter(annotationClass -> getReturnTypeOfAnnotationMethod((Class<? extends Annotation>) annotationClass,
+                RegisterValidator.REQUEST_TYPE_METHOD_NAME)
+                .equals(requestType))
+            .filter(annotationClass -> allowedVersionTypes.contains(getReturnTypeOfAnnotationMethod(
+                (Class<? extends Annotation>) annotationClass,
+                    RegisterValidator.MAX_VERSION_METHOD_NAME)))
+            .map(annotationClass -> (Class<? extends Annotation>) annotationClass)
```

Comment:
Should we filter here on the processingPhase method? I believe so, but as there is no filtering for that in the tests either I am starting to have a doubt, but if we should not, can you please elaborate why?

Reply:
We are checking it here:
https://github.com/apache/ozone/blob/a39467057846bcf669380289f5674bb6eed9e752/hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/request/validation/ValidatorRegistry.java#L251
We shouldn't silently filter out methods, without the logs we might not be able to figure out what got skipped.